File: new.c
#include <assert.h>
#include <stdarg.h>
#include <stdlib.h>

#include "new.r"

void	*new(const void *_class, ...)
{
	const struct s_Class	*class = _class;
	void					*p;
	va_list					ap;

	p = calloc(1, class->size);
	assert(p);
	*(const struct s_Class **)p = class;
	if (class->ctor)
	{
		va_start(ap, _class);
		p = class->ctor(p, &ap);
		va_end(ap);
	}
	return (p);
}

void	delete(void *self)
{
	const struct s_Class	**cp = self;

	if (self && *cp && (*cp)->dtor)
		self = (*cp)->dtor(self);
	free(self);
}

void	draw(const void *self)
{
	const struct s_Class *const	*cp = self;

	if (self && *cp && (*cp)->draw)
		(*cp)->draw(self);
}

size_t	sizeOf(const void *self)
{
	const struct s_Class *const	*cp = self;

	assert(self && *cp);
	return ((*cp)->size);
}

File: new.r
#include <stdarg.h>
#include <stdlib.h>

struct s_Object
{
	unsigned int	count;
	struct s_Set	*in;
};

struct s_Class
{
	size_t	size;
	void	*(*ctor)(void *self, va_list *app);
	void	*(*dtor)(void *self);
	void	(*draw)(const void *self);
};

struct s_Set {
	unsigned int	count;
};
